src/routes/projects.rs

```diff
@@ -27,3 +32,31 @@ pub fn create(
         token: token.to_string()?,
     }))
 }
+
+#[post("/images", data = "<data>")]
+pub fn upload_image(
+    conn: RequestContext,
+    content_type: &ContentType,
+    token: Token,
+    data: Data,
+) -> core::result::Result<Json<MediumResponse>, Custom<String>> {
+    let boundary = validate_medium_upload(&conn, content_type, &token)?;
+    let entries = process_file_upload(boundary, data)?;
+    let project_id = get_foreign_key("project_id", &entries)?;
+    let database_context = match conn.database_context(
+        Resource::Medium,
+        Some(&token),
+        ActionType::Create,
+        ActionModifier::Own,
+    ) {
+        Ok(ctx) => ctx,
+        Err(err) => return Err(Custom(Status::Unauthorized, err.to_string())),
```

Comment:
Could you use a `map_err` here? Something like

```rust
conn.database_context(
        Resource::Medium,
        Some(&token),
        ActionType::Create,
        ActionModifier::Own,
    ).map_err(|err| Custom(Status::Unauthorized, err.to_string()) )?;
```

Reply:
Seems like a good idea. In general I want to figure out a better way of autoconverting errors. Ideally something that automatically takes any failure::Error and converts to Custom

Reply:
What exactly does `Custom` do?

Reply:
It's a rocket builtin that allows you to return specific error codes

Reply:
I.e. 404/401/402

Reply:
Or 418, the best error code

Reply:
Maybe we should just impl upstream then? Or create a utility function?

```rust
pub fn custom_http_error<T, E>(error: Result<T, E>, code: u16) -> Result<T, Custom>
where
  E: std::string::ToString,
{
  error.map_err(|e| { Custom(code, error.to_string()) })
}
```

Or our own fun version of the `?` operator, since we're already on the nightly version of Rust.

Reply:
Okay, no, we're not making our own `try!`. I vote that we make errors somewhat systematic and maybe include an error chain (fail-chain?). Ideally in a JSON format that our front end can parse and display. The issue with blindly doing `to_string` is that we end up with errors that aren't JSON but just some random string.

Reply:
What do you mean by that? I'm not sure what you mean by "systematic" or what you have in mind with the phrase "error chain".

Reply:
Oh [error chain](https://github.com/rust-lang-nursery/error-chain) is a crate. And systematic basically means we make the error type consistent across the app. Right now half of the routes are using Custom while the other half are using `failure::Error`. Also, if we can make error response schema consistent, i.e. we always return an `{ message: String, trace: String }` then that would be fantastic